lib/lnd_client.ex

```diff
@@ -18,12 +21,41 @@ defmodule LndClient do
 
   @long_timeout 500_000
 
-  def start() do
-    GenServer.start(__MODULE__, %{} |> init_subscriptions, name: __MODULE__)
+  def start(%ConnConfig{} = conn_config) do
+    GenServer.start(__MODULE__, init_state(conn_config), name: __MODULE__)
   end
 
-  def start_link(_) do
-    GenServer.start_link(__MODULE__, %{} |> init_subscriptions, name: __MODULE__)
+  @doc """
+  Starts a process which connects to an LND instance.
+
+  ## Examples
+
+      iex> {:ok, pid} = %LndClient.Config.new{
+        node_uri: "127.0.0.1:10001",
+        cert_path: "/path/to/tls.cert",
+        macaroon_path: "/path/to/readonly.macaroon"
+      } |> LndClient.start_link()
+
+  You may pass something like the following to a Supervisor:
+
+  ```ex
+  {
+    LndClient,
+    %LndClient.ConnConfig{
+      node_uri: System.get_env("NODE_URI"),
+      cert_path: System.get_env("CERT_PATH"),
+      macaroon_path: System.get_env("MACAROON_PATH")
+    },
+  },
+  ```
+  """
+  def start_link(%ConnConfig{} = conn_config) do
+    case GenServer.start_link(__MODULE__, init_state(conn_config), name: __MODULE__) do
```

Comment:
I just noticed that the connect function makes a call to the GenServer before it's even been initiated...

I think the connect function should be a normal function, but called from within the `init` function, and thus in the context of the GenServer itself.

Reply:
Ah yes, that's right! Because in the init the server would have started already

Reply:
I can't seem to call `connect()` from the `init` function. Is it because it hasn't been started yet and therefore this exception is raised?

```
17:00:15.371 [error] Process #PID<0.77.0> raised an exception
** (ArgumentError) errors were found at the given arguments:

  * 2nd argument: not a pid

    :erlang.monitor(:process, {:error, {:calling_self, {GenServer, :call, [LndClient, :connect, :infinity]}}})
    (iex 1.14.0) lib/iex/server.ex:100: IEx.Server.run_without_registration/1
```

Reply:
Ah, well, scratch that. I got something working. What worked was to not make a call to the genserver, but rather connect already and return that state in `init`.

```ex
  def init(state) do
    connect(state) # call the Connectivity.connect code here
  end
```

The thing is this may return `{:error, ...}`. I haven't seen `init` ever return those in examples online.

Reply:
Yeah, `start_link` is in the process of the caller, but `init` is executing in the GenServer's own process. When calling from outside, you got to switch context, thus make calls or casts, but from init, all you got to do is to do the job straight away.